### ISSUE TITLE
New rule: Reports the usages of `<expr1>?.let { <expr2> } ?: run { <single_line_expr> }`

### ISSUE THREAD

**Issue by atulgpt** [detekt/detekt#6112]
## Expected Behavior of the rule
Rule should report `<expr1>?.let { <expr2> } ?: run { <single_line_expr> }` as expression can be simplified as `<expr1>?.let { <expr2> } ?: <single_line_expr>`. Here there is an unnecessary use of `run { }` block

uncompliant code
```
val nullable: Int? = 4
val default = 1
val a = nullable?.let { 0 } ?: run { default }
```

compliant code
```
val nullable: Int? = 4
val default = 1
val a = nullable?.let { 0 } ?: default
```

## Context
Got this review in https://github.com/detekt/detekt/pull/5981#discussion_r1197842827 having unnecessary `run { }` really makes the code more indented and increases the complexity

**Comment by 3flex** [comment]
Could the rule be more generic?

It looks like what we want flagged is any use of the non-extension version of `run` that contains a single line expression.

**Comment by atulgpt** [comment]
Yes, I think It can be generic. But then we also have to consider single expression use of `with`, `apply`, `let` and if those make sense as well or not(and then we can maybe incorporate more scoping operators). Let me update the proposal

**Comment by atulgpt** [comment]
Hi, @3flex got your query yes it should be generic.

We can create similar rules for `with` as well as `apply` as well
Some more violating code
```
val length = with("") {
       length
}
```
```
val length = "".apply {
       length
}
```
```
val a = run { "".length }
```

**Comment by BraisGabin** [comment]
So the proposal would be to create `UnnecessaryRun` and `UnnecessaryWith` too? I would vote in favor of that. And these will not be related at all with the previous `<expr1>?.let { <expr2> } ?:`, right?

If that's the idea, we need proper definitions of what these rules should catch. Some more examples of compilant/non-compilant.

**Comment by atulgpt** [comment]
Yes @BraisGabin this suggestion is independent of the previous suggestion as that one deals through the case when due to the last nullable expression code execution pass through the next `run` or `apply` block as well.
This one mostly focuses on unnecessary block operators when there is only one expression

**Comment by atulgpt** [comment]
Not stale

**Comment by BraisGabin** [comment]
Sorry for the really late response. How can we make this issue actionable? Would a rule called `UnnecessaryRun` that detects single expresions inside `run` close this issue?

**Comment by atulgpt** [comment]
Yes, @BraisGabin that will close the issue. After implementing that and analyzing the learnings from that rule we can implement more rules mentioned in https://github.com/detekt/detekt/issues/6112#issuecomment-1563574998

**Comment by BraisGabin** [comment]
Ok, so to summarize to anyone reading and willing to contribute. We want a rule called `UnnecessaryRun` that detects single expressions. And we are only talking about `run` without any receiver. `T.run` is another topic.

This rule can be expanded in the future but for now just that.

**Comment by exijn** [comment]
Hi, I'd like to work on this issue.

I'll follow the final scope described in the discussion and implement an `UnnecessaryRun` rule that reports a receiverless `run` block containing a single expression. Receiver-based calls such as `value.run { ... }` and blocks containing multiple statements will remain out of scope.

I'll also register the rule, add the required documentation and configuration entry, and cover compliant and non-compliant cases with tests.
